java/src/main/c++/jni_spark_vw.cc

```diff
@@ -384,24 +384,14 @@ JNIEXPORT void JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitExample_clear(JNI
   }
 }
 
-void addNamespaceIfNotExists(vw* all, example* ex, char ns)
-{
-  if (std::find(ex->indices.begin(), ex->indices.end(), ns) == ex->indices.end())
-  {
-    ex->indices.push_back(ns);
-  }
-}
-
 JNIEXPORT void JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitExample_addToNamespaceDense(
     JNIEnv* env, jobject exampleObj, jchar ns, jint weight_index_base, jdoubleArray values)
 {
   INIT_VARS
 
   try
   {
-    addNamespaceIfNotExists(all, ex, ns);
-
-    auto features = ex->feature_space.data() + ns;
+    auto& features = ex->feature_space.get_or_create_feature_group(weight_index_base, ns);

```

Comment:
How does this work?  Just gets the first feature group that matches ns?

Reply:
To create a namespace you must know both the hash and index for it, so the creation function need the complete info for creation. Duplicates are not allowed, so if the hash already exists you get back that existing feature_group